<a href="https://colab.research.google.com/github/iqra-amer/Flyrank-ML-Internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 1 – Research Question Framing

**Lane:** Refresh / Content Opportunity Scoring

## 1. My Lane and Why

### Selected Lane

**Refresh / Content Opportunity Scoring**

### Why I Chose This Lane

I chose the **Refresh / Content Opportunity Scoring** lane because organizations often manage thousands of content pages, making it impractical to review each page manually. A systematic approach that ranks pages based on available data can help prioritize which pages should be reviewed first. I am interested in understanding how data and machine learning can support this decision-making process by providing evidence-based recommendations rather than relying only on manual judgment.

## 2. The Research Question

### Research Question

How can available content performance data be used to prioritize webpages that should be reviewed for content refresh?

### Decision

The decision is to determine which webpages should be placed at the top of the review queue for content updates.

### Action

A content manager can review the highest-ranked pages first, update outdated or underperforming content, and monitor whether those updates improve content performance.

### Cost of a Wrong Recommendation

If important pages are ranked too low, valuable opportunities to improve content performance may be missed. If pages that do not need attention are ranked too high, time and resources may be spent on unnecessary updates.

### Why Data or Machine Learning Can Help

Manually reviewing thousands of webpages is time-consuming and subjective. Data can help identify useful patterns, while machine learning may assist in ranking pages based on multiple signals to support more consistent and evidence-based decisions.

In [1]:
import os
import sys
import subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)

print("Repository is ready!")

Repository is ready!


In [6]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Dataset shape:", df.shape)

df.head()


Dataset shape: (30000, 44)


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [11]:
# Number of pages in the dataset
print("Total number of pages:", len(df))

Total number of pages: 30000


### Initial Observations

The starter dataset contains **30,000 webpages**.

The trend distribution shows that **16,262 pages are currently declining**, making it the largest category in the dataset. In comparison, **4,388 pages are improving**, while **5,962 pages are stable**.

These observations suggest that manually reviewing every declining page would be impractical. This supports the choice of the **Refresh / Content Opportunity Scoring** lane, where data-driven methods may help prioritize which pages should be reviewed first.


In [12]:
# Counting the trend
df["trend_direction"].value_counts()

,count
trend_direction,
down,16262
stable,5962
up,4388
new,2236
flat,1152


In [9]:
df["age_tier"].value_counts()

,count
age_tier,
91-180,11780
181-365,11368
365+,6360
31-90,492


In [10]:
df["impressions_90d"].describe()

,impressions_90d
count,30000.000000
mean,5200.366300
std,16838.019547
min,1.000000
25%,81.000000
50%,731.000000
75%,3615.250000
max,517715.000000


# 3. Quick Look at the Data

The starter dataset contains **30,000 webpages**.

Several observations support choosing the Refresh / Content Opportunity Scoring lane.

- The largest trend category is **down (16,262 pages)**, suggesting many pages are experiencing declining performance.
- There are **6,360 pages** that are more than one year old (365+ days).
- The average number of impressions is approximately **5,200**, while the median is only **731**, indicating that a relatively small number of pages receive extremely high traffic compared with most pages.

These observations suggest that reviewing every page manually would not be practical. A ranking system could help prioritize which pages deserve attention first.

# 4. Careful Words: What I Can and Cannot Claim

This notebook does not demonstrate that refreshing content improves search performance.

It also does not claim that every declining or older page should be updated.

The analysis only shows that observable signals exist which may help prioritize manual review.

Any future model should be treated as a decision-support tool rather than an automated decision maker.

# 5. Self-Check

- [x] Selected one predefined lane.
- [x] Clearly stated the research question.
- [x] Identified the unit of analysis.
- [x] Explained the decision and action.
- [x] Explained the cost of an incorrect recommendation.
- [x] Included supporting observations from the starter dataset.
- [x] Used careful language without making unsupported claims.